lib/teleterm/daemon/daemon_headless.go

```diff
@@ -185,6 +182,73 @@ func (s *Service) startHeadlessWatcher(cluster *clusters.Cluster) error {
 	return nil
 }
 
+// sendPendingHeadlessAuthentication notifies the Electron App of a pending headless authentication.
+func (s *Service) sendPendingHeadlessAuthentication(ctx context.Context, ha *types.HeadlessAuthentication, cluster *clusters.Cluster) error {
+	// headless authentication requests will timeout after 3 minutes, so we can close the
+	// Electron modal once this time is up.
+	sendCtx, cancel := context.WithTimeout(ctx, defaults.CallbackTimeout)
+	defer cancel()
+
+	req := &api.SendPendingHeadlessAuthenticationRequest{
+		RootClusterUri:                 cluster.URI.String(),
+		HeadlessAuthenticationId:       ha.GetName(),
+		HeadlessAuthenticationClientIp: ha.ClientIpAddress,
+	}
+
+	// cancel the request early if the headless request is resolved or deleted.
+	go func() {
+		err := waitForHeadlessAuthenticationResolution(sendCtx, ha, cluster)
+		if err != nil {
+			s.cfg.Log.WithField("cluster", cluster.URI.String()).WithError(err).Debug("Failed to wait for headless authentication resolution")
```

Comment:
Since `waitForHeadlessAuthenticationResolution` creates a client from scratch and has to go through the handshake, it's possible that the headless request is resolved before we create a watcher.

But even if we first initialized a watcher, then send a request to tshdEventsClient and then started listening for the update, we're still not guaranteed that the request wasn't approved somewhere else.

Since it has potential for some weird edge cases, could we add a field to this log with the headless authn UUID?

Reply:
Originally I planned to ignore this edge case, but I'll try fixing it by sharing a watcher among pending headless requests.